File: packages/nodejs/whatsapp-sdk/tests/client.test.ts
import { describe, expect, it } from "vitest";
import { CobrancaQuantityController } from "../src/agenda.js";
import { WhatsAppClient } from "../src/client.js";
import { computeCobrancaQuantity } from "../src/cobranca-quantity.js";
import { StaticEntitlementChecker } from "../src/entitlement.js";
import { EntitlementError, QuotaExceededError } from "../src/errors.js";
import { InMemoryUsageStore } from "../src/metering.js";
import { CENARIO_BASE_PLAN } from "../src/plan.js";
import type {
  AgendaCobrancaLike,
  CobrancaRecord,
  CreateCobrancaInput,
  FetchLike,
  ListCobrancasInput,
} from "../src/types.js";

interface Captured {
  request: Request;
  body: string;
}

function mockFetch(
  handler: (req: Request, body: string) => Response | Promise<Response>,
): { fetch: FetchLike; calls: () => Captured[] } {
  const calls: Captured[] = [];
  const fetchImpl: FetchLike = (async (input, init) => {
    const request = new Request(input, init);
    const body = await request.clone().text();
    calls.push({ request, body });
    return handler(request, body);
  }) as FetchLike;
  return { fetch: fetchImpl, calls: () => calls };
}

function jsonResponse(status: number, body: unknown): Response {
  return new Response(JSON.stringify(body), {
    status,
    headers: { "Content-Type": "application/json" },
  });
}

const META_OK = {
  messaging_product: "whatsapp",
  contacts: [{ input: "5511999999999", wa_id: "5511999999999" }],
  messages: [{ id: "wamid.TEST" }],
};

const PAGADOR = { documento: "12345678901", nome: "Empresa SaaS", email: "fin@example.com" };

class MemoryAgenda implements AgendaCobrancaLike {
  readonly items = new Map<string, CobrancaRecord>();
  readonly cobrancas = {
    create: async (input: CreateCobrancaInput): Promise<CobrancaRecord> => {
      const rec: CobrancaRecord = {
        id: `cob-${input.externalReference}`,
        externalReference: input.externalReference,
        valorCentavos: input.valorCentavos,
        vencimento: input.vencimento,
        status: "pendente",
        pagador: input.pagador,
      };
      this.items.set(rec.id, rec);
      return rec;
    },
    list: async (filtros: ListCobrancasInput = {}) => {
      const data = [...this.items.values()].filter((c) => {
        if (filtros.externalReference && c.externalReference !== filtros.externalReference) return false;
        return true;
      });
      return { data };
    },
    cancel: async (id: string) => {
      const rec = this.items.get(id);
      if (!rec) throw new Error("missing");
      rec.status = "cancelada";
      return rec;
    },
  };
}

describe("WhatsAppClient", () => {
  it("rejeita configuracao incompleta", () => {
    expect(() => new WhatsAppClient({ accessToken: "", phoneNumberId: "", tenantId: "" })).toThrow(
      /accessToken/,
    );
  });

  it("bloqueia envio sem add-on habilitado", async () => {
    const { fetch } = mockFetch(() => jsonResponse(200, META_OK));
    const client = new WhatsAppClient({
      accessToken: "token",
      phoneNumberId: "123",
      tenantId: "acme",
      entitlement: new StaticEntitlementChecker(false, { message: "addon off" }),
      fetch,
    });
    await expect(
      client.sendText({ to: "5511999999999", body: "oi" }),
    ).rejects.toBeInstanceOf(EntitlementError);
  });

  it("envia template utility, marca categoria e mede uso", async () => {
    const { fetch, calls } = mockFetch(() => jsonResponse(200, META_OK));
    const client = new WhatsAppClient({
      accessToken: "token",
      phoneNumberId: "555",
      tenantId: "acme",
      entitlement: new StaticEntitlementChecker(true),
      clock: () => new Date("2026-09-16T15:00:00Z"),
      fetch,
    });

    const result = await client.sendTemplate({
      to: "5511999999999",
      category: "utility",
      template: { name: "pix_recebido", language: "pt_BR" },
    });

    expect(result.category).toBe("utility");
    expect(result.meta.messages?.[0]?.id).toBe("wamid.TEST");
    expect(result.metering.used).toBe(1);
    expect(result.metering.billableDelta).toBe(1);
    expect(result.cobrancaPlan?.totalQuantity).toBe(1);

    const captured = calls()[0];
    expect(captured.request.method).toBe("POST");
    expect(new URL(captured.request.url).pathname).toBe("/v21.0/555/messages");
    expect(captured.request.headers.get("Authorization")).toBe("Bearer token");
    const body = JSON.parse(captured.body);
    expect(body.type).toBe("template");
    expect(body.template.name).toBe("pix_recebido");
    expect(JSON.parse(body.biz_opaque_callback_data)).toEqual({ category: "utility" });
    expect(captured.body).not.toContain("token");
  });

  it("faz rollback da cota se a Graph API falhar", async () => {
    const { fetch } = mockFetch(() =>
      jsonResponse(400, { error: { message: "invalid to", type: "OAuthException", code: 100 } }),
    );
    const store = new InMemoryUsageStore();
    const client = new WhatsAppClient({
      accessToken: "token",
      phoneNumberId: "555",
      tenantId: "acme",
      entitlement: new StaticEntitlementChecker(true),
      usageStore: store,
      fetch,
    });
    await expect(client.sendText({ to: "1", body: "oi" })).rejects.toMatchObject({ status: 400 });
    const snap = await store.get("acme", client.metering.period());
    expect(snap.counts.service).toBe(0);
  });

  it("modo hard impede o POST a Meta", async () => {
    let graphCalls = 0;
    const { fetch } = mockFetch(() => {
      graphCalls += 1;
      return jsonResponse(200, META_OK);
    });
    const client = new WhatsAppClient({
      accessToken: "token",
      phoneNumberId: "555",
      tenantId: "acme",
      entitlement: new StaticEntitlementChecker(true),
      quotaMode: "hard",
      plan: { categories: { service: { monthlyQuota: 1 } } },
      fetch,
    });
    await client.sendText({ to: "5511", body: "1" });
    await expect(client.sendText({ to: "5511", body: "2" })).rejects.toBeInstanceOf(QuotaExceededError);
    expect(graphCalls).toBe(1);
  });
});

describe("CobrancaQuantityController", () => {
  it("cria cobrancas do plano e cancela extras quando o uso cai", async () => {
    const agenda = new MemoryAgenda();
    const controller = new CobrancaQuantityController(agenda, PAGADOR, 10);

    const full = computeCobrancaQuantity(
      { auth: 20_000, utility: 0, service: 0, marketing: 0 },
      CENARIO_BASE_PLAN,
      { period: "2026-09", tenantId: "acme", strategy: "by-sessions", sessionsPerCobranca: 10_000 },
    );
    expect(full.totalQuantity).toBe(2);

    const first = await controller.sync(full);
    expect(first.created).toBe(2);
    expect(first.cancelled).toBe(0);
    expect([...agenda.items.values()].filter((c) => c.status === "pendente")).toHaveLength(2);
    expect(first.items[0]?.cobranca?.vencimento).toBe("2026-09-10");

    const reduced = computeCobrancaQuantity(
      { auth: 10_000, utility: 0, service: 0, marketing: 0 },
      CENARIO_BASE_PLAN,
      { period: "2026-09", tenantId: "acme", strategy: "by-sessions", sessionsPerCobranca: 10_000 },
    );
    const second = await controller.sync(reduced);
    expect(second.kept).toBe(1);
    expect(second.cancelled).toBe(1);
    expect(second.created).toBe(0);
    const pending = [...agenda.items.values()].filter((c) => c.status === "pendente");
    expect(pending).toHaveLength(1);
    expect(pending[0]?.externalReference).toBe("wa:acme:2026-09:auth:1");
  });
});

File: packages/go/whatsapp-go/client_test.go
package whatsapp

import (
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync"
	"testing"
	"time"
)

type captured struct {
	Method string
	Path   string
	Auth   string
	Body   string
}

func graphServer(t *testing.T, status int, payload any, calls *[]captured) *httptest.Server {
	t.Helper()
	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		b, _ := io.ReadAll(r.Body)
		*calls = append(*calls, captured{Method: r.Method, Path: r.URL.Path, Auth: r.Header.Get("Authorization"), Body: string(b)})
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(status)
		_ = json.NewEncoder(w).Encode(payload)
	}))
}

var metaOK = map[string]any{
	"messaging_product": "whatsapp",
	"contacts":          []any{map[string]any{"input": "5511999999999", "wa_id": "5511999999999"}},
	"messages":          []any{map[string]any{"id": "wamid.TEST"}},
}

func TestClientConfigIncompleta(t *testing.T) {
	_, err := NewClient(Options{})
	if err == nil || !strings.Contains(err.Error(), "AccessToken") {
		t.Fatalf("%v", err)
	}
}

func TestClientBloqueiaSemAddon(t *testing.T) {
	var calls []captured
	srv := graphServer(t, 200, metaOK, &calls)
	defer srv.Close()
	client, err := NewClient(Options{
		AccessToken:   "token",
		PhoneNumberID: "123",
		TenantID:      "acme",
		Entitlement:   StaticEntitlementChecker{Enabled: false, Message: "addon off"},
		GraphBaseURL:  srv.URL,
		HTTPClient:    srv.Client(),
	})
	if err != nil {
		t.Fatal(err)
	}
	_, err = client.SendText("5511", "oi", CategoryService, "")
	if _, ok := err.(*EntitlementError); !ok {
		t.Fatalf("%v", err)
	}
}

func TestSendTemplateUtility(t *testing.T) {
	var calls []captured
	srv := graphServer(t, 200, metaOK, &calls)
	defer srv.Close()
	client, err := NewClient(Options{
		AccessToken:   "token",
		PhoneNumberID: "555",
		TenantID:      "acme",
		Entitlement:   StaticEntitlementChecker{Enabled: true},
		Clock:         func() time.Time { return time.Date(2026, 9, 16, 15, 0, 0, 0, time.UTC) },
		GraphBaseURL:  srv.URL,
		GraphVersion:  "v21.0",
		HTTPClient:    srv.Client(),
	})
	if err != nil {
		t.Fatal(err)
	}
	result, err := client.SendTemplate("5511999999999", CategoryUtility, TemplatePayload{Name: "pix_recebido", Language: "pt_BR"}, "")
	if err != nil {
		t.Fatal(err)
	}
	if result.Category != CategoryUtility || result.Meta.Messages[0].ID != "wamid.TEST" {
		t.Fatalf("%+v", result)
	}
	if result.Metering.Used != 1 || result.CobrancaPlan.TotalQuantity != 1 {
		t.Fatalf("metering %+v plan %+v", result.Metering, result.CobrancaPlan)
	}
	if len(calls) != 1 || calls[0].Method != http.MethodPost {
		t.Fatalf("%+v", calls)
	}
	if !strings.HasSuffix(calls[0].Path, "/v21.0/555/messages") {
		t.Fatalf("path %s", calls[0].Path)
	}
	if calls[0].Auth != "Bearer token" {
		t.Fatal(calls[0].Auth)
	}
	var body map[string]any
	_ = json.Unmarshal([]byte(calls[0].Body), &body)
	if body["type"] != "template" {
		t.Fatalf("%v", body)
	}
	tpl := body["template"].(map[string]any)
	if tpl["name"] != "pix_recebido" {
		t.Fatal(tpl)
	}
	var cb map[string]any
	_ = json.Unmarshal([]byte(body["biz_opaque_callback_data"].(string)), &cb)
	if cb["category"] != "utility" {
		t.Fatal(cb)
	}
}

func TestRollbackSeGraphFalhar(t *testing.T) {
	var calls []captured
	srv := graphServer(t, 400, map[string]any{"error": map[string]any{"message": "invalid to"}}, &calls)
	defer srv.Close()
	store := NewInMemoryUsageStore()
	client, err := NewClient(Options{
		AccessToken:   "token",
		PhoneNumberID: "555",
		TenantID:      "acme",
		Entitlement:   StaticEntitlementChecker{Enabled: true},
		Store:         store,
		GraphBaseURL:  srv.URL,
		HTTPClient:    srv.Client(),
	})
	if err != nil {
		t.Fatal(err)
	}
	_, err = client.SendText("1", "oi", CategoryService, "")
	if err == nil {
		t.Fatal("esperado erro")
	}
	snap, _ := store.Get("acme", client.Metering.Period())
	if snap.Counts.Service != 0 {
		t.Fatalf("nao fez rollback: %d", snap.Counts.Service)
	}
}

func TestHardNaoChamaMeta(t *testing.T) {
	var calls []captured
	srv := graphServer(t, 200, metaOK, &calls)
	defer srv.Close()
	quota := int64(1)
	client, err := NewClient(Options{
		AccessToken:   "token",
		PhoneNumberID: "555",
		TenantID:      "acme",
		Entitlement:   StaticEntitlementChecker{Enabled: true},
		QuotaMode:     QuotaHard,
		Plan: &PlanOverride{Categories: map[MessageCategory]CategoryQuotaPatch{
			CategoryService: {MonthlyQuota: &quota},
		}},
		GraphBaseURL: srv.URL,
		HTTPClient:   srv.Client(),
	})
	if err != nil {
		t.Fatal(err)
	}
	if _, err := client.SendText("5511", "1", CategoryService, ""); err != nil {
		t.Fatal(err)
	}
	_, err = client.SendText("5511", "2", CategoryService, "")
	if _, ok := err.(*QuotaExceededError); !ok {
		t.Fatalf("%v", err)
	}
	if len(calls) != 1 {
		t.Fatalf("graph calls %d", len(calls))
	}
}

type memoryAgenda struct {
	mu    sync.Mutex
	items map[string]CobrancaRecord
}

func newMemoryAgenda() *memoryAgenda {
	return &memoryAgenda{items: map[string]CobrancaRecord{}}
}

func (m *memoryAgenda) CreateCobranca(in CreateCobrancaInput) (CobrancaRecord, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	rec := CobrancaRecord{
		ID: "cob-" + in.ExternalReference, ExternalReference: in.ExternalReference,
		ValorCentavos: in.ValorCentavos, Vencimento: in.Vencimento, Status: "pendente", Pagador: &in.Pagador,
	}
	m.items[rec.ID] = rec
	return rec, nil
}

func (m *memoryAgenda) ListCobrancas(in ListCobrancasInput) ([]CobrancaRecord, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	var out []CobrancaRecord
	for _, c := range m.items {
		if in.ExternalReference != "" && c.ExternalReference != in.ExternalReference {
			continue
		}
		out = append(out, c)
	}
	return out, nil
}

func (m *memoryAgenda) CancelCobranca(id string) (CobrancaRecord, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	rec := m.items[id]
	rec.Status = "cancelada"
	m.items[id] = rec
	return rec, nil
}

func TestSyncCriaECancela(t *testing.T) {
	agenda := newMemoryAgenda()
	ctrl := NewCobrancaQuantityController(agenda, Pagador{Documento: "12345678901", Nome: "Empresa SaaS"}, 10)
	full := ComputeCobrancaQuantity(UsageCounts{Auth: 20_000}, CenarioBasePlan, ConversionOptions{
		Period: "2026-09", TenantID: "acme", Strategy: StrategyBySessions, SessionsPerCobranca: 10_000,
	})
	if full.TotalQuantity != 2 {
		t.Fatalf("qty %d", full.TotalQuantity)
	}
	first, err := ctrl.Sync(full)
	if err != nil {
		t.Fatal(err)
	}
	if first.Created != 2 || first.Cancelled != 0 {
		t.Fatalf("%+v", first)
	}
	if first.Items[0].Cobranca.Vencimento != "2026-09-10" {
		t.Fatal(first.Items[0].Cobranca.Vencimento)
	}
	reduced := ComputeCobrancaQuantity(UsageCounts{Auth: 10_000}, CenarioBasePlan, ConversionOptions{
		Period: "2026-09", TenantID: "acme", Strategy: StrategyBySessions, SessionsPerCobranca: 10_000,
	})
	second, err := ctrl.Sync(reduced)
	if err != nil {
		t.Fatal(err)
	}
	if second.Kept != 1 || second.Cancelled != 1 || second.Created != 0 {
		t.Fatalf("%+v", second)
	}
	pending := 0
	var ref string
	for _, c := range agenda.items {
		if c.Status == "pendente" {
			pending++
			ref = c.ExternalReference
		}
	}
	if pending != 1 || ref != "wa:acme:2026-09:auth:1" {
		t.Fatalf("pending %d ref %s", pending, ref)
	}
}
